extensions/resteasy-reactive/quarkus-resteasy-reactive/runtime/src/main/java/io/quarkus/resteasy/reactive/server/runtime/ResteasyReactiveRecorder.java

```diff
@@ -212,13 +212,48 @@ public void accept(RoutingContext routingContext) {
         return new ResteasyReactiveVertxHandler(eventCustomizer, initialHandler);
     }
 
-    public Handler<RoutingContext> failureHandler(RuntimeValue<RestInitialHandler> restInitialHandlerRuntimeValue) {
+    public Handler<RoutingContext> failureHandler(RuntimeValue<RestInitialHandler> restInitialHandlerRuntimeValue,
+            boolean noCustomAuthCompletionExMapper, boolean noCustomAuthFailureExMapper, boolean noCustomAuthRedirectExMapper,
+            boolean proactive) {
         final RestInitialHandler restInitialHandler = restInitialHandlerRuntimeValue.getValue();
         // process auth failures with abort handlers
         return new Handler<RoutingContext>() {
             @Override
             public void handle(RoutingContext event) {
 
+                // special handling when proactive auth is enabled as then we know default auth failure handler already run
+                if (proactive && event.get(QuarkusHttpUser.AUTH_FAILURE_HANDLER) instanceof DefaultAuthFailureHandler) {
+                    // we want to prevent repeated handling of exceptions if user don't want to handle exception himself
+                    // we do not pass exception to abort handlers if proactive auth is enabled and user did not
+                    // provide custom ex. mapper; we replace default auth failure handler as soon as we can, so that
+                    // we can handle Quarkus Security Exceptions ourselves
+                    if (event.failure() instanceof AuthenticationFailedException) {
```

Comment:
Can you please consider, later, add a base `AuthenticationException` which these 3 exceptions will extend, to `quarkus-security`, it can simplify the code as in a few places we check all of these 3 exceptions and more can be added

Reply:
that's clever, I didn't thought about it :-(, I'll look into it later (busy ATM).

Reply:
@michalvavrik This is a limit of my Java related optimization ideas :-). Sure, later, I mean next year some time, really no rush at all :-)